### ISSUE TITLE
pass on cookies from server-side API calls on SSR response

### ISSUE THREAD

**Issue by antoineanorra** [nuxt/framework#2479]
### Environment

------------------------------
- Operating System: `Linux`
- Node Version:     `v16.13.1`
- Nuxt Version:     `3.0.0-27333874.279bfdc`
- Package Manager:  `yarn@1.22.17`
- Bundler:          `Vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

Currently no reproduction link is up to date with the latest version of Nuxt3 so I can't help you

### Describe the bug

Hello Nuxt 3 !

I have a problem, I try to make a session manager with redis (like express-session) but here I will try to make it in h3/nitro version, I don't know if it is possible ?

`setCookie` of h3 does not send the cookie to the client during the server side rendering.
However, it does send it if a request is executed directly on the api path.

server/api/hello.js
```js
import { setCookie } from 'h3'

export default (req, res) => {
	setCookie(res, "sessionId", "new session id");
	return 'you are now connected ! (in reality not)';
}
```

hello.vue
```js
const { data } = await useFetch("/api/hello", {
	headers: useRequestHeaders(),
});
```

Browser that executes the request to the API in SSR :
![image](https://user-images.githubusercontent.com/48593504/146915926-214442d7-6053-4205-bbfd-9f09513951c6.png)

Directly by making a request to the API with Postman :
![image](https://user-images.githubusercontent.com/48593504/146916952-6b737fba-5114-4e29-a1b2-d75d09ed9aa3.png)

Thanks !

**Comment by Atinux** [comment]
I created a small sandbox, maybe it could help @pi0 @danielroe 

https://stackblitz.com/edit/nuxt-starter-zi9z7u?file=pages%2Findex.vue

**Comment by danielroe** [comment]
In a web browser, any of many requests can set cookies for the browser. However, making requests on the server results in different behaviour. If you want to allow the headers set on the response to `useFetch` to be added to the response from SSR, you'll have to handle that yourself. (You could even write your own composable to do this.)

If we proxy headers (even just cookie headers) from _server_ API requests to the SSR response by default, that risks leaking sensitive information.

**Comment by antoineanorra** [comment]
`If we proxy headers (even just cookie headers) from server API requests to the SSR response by default, that risks leaking sensitive information.`

@danielroe Is it possible to have a boolean configuration to allow this?

**Comment by hale01** [comment]
> In a web browser, any of many requests can set cookies for the browser. However, making requests on the server results in different behaviour. If you want to allow the headers set on the response to `useFetch` to be added to the response from SSR, you'll have to handle that yourself. (You could even write your own composable to do this.)
> 
> If we proxy headers (even just cookie headers) from _server_ API requests to the SSR response by default, that risks leaking sensitive information.

I have the same problem.

useFetch on ssr works fine.
But when useFetch is used on the client, no cookies are sent.

Prescribed them explicitly by hand in the request, there is no effect.

```javascript
// Fine SSR. On client cookie not sent.
const { data, error } = await useFetch(fetchUrl, { 
  baseURL: $config.baseUrl,
  headers: {cookie: 'csrftoken=tokenHere; sessionid=sessionidHere;' },
})
```

**Comment by hale01** [comment]
> I created a small sandbox, maybe it could help @pi0 @danielroe
> 
> https://stackblitz.com/edit/nuxt-starter-zi9z7u?file=pages%2Findex.vue

The solution has the right to life, but passing cookies in GET parameters is not beautiful from an aesthetic point of view and complicates the code.

There are also httponly cookies that must always be sent, but we cannot read them.

**Comment by antoineanorra** [comment]
Someone else is talking about it here, let me repost what I said in the discussion : #2925 

> The problem is that a useFetch call on the browser side is required for Nuxt/Nitro to accept to deliver the cookie to the browser. This means that on each SSR visit, a "new" client cannot be given a cookie, so it is absolutely necessary to create a no-ssr call for the client to get its cookie. This implies that for each new SSR visitor, a cookie is issued and information is created in Redis/database that will never be used because the browser will never receive it

**Comment by danielroe** [comment]
I think there might be some misunderstanding here about the flow of cookies.

1. Initial request (**REQ1** and **RES1**) comes in from client, arrives at our Nitro server.
2. Whilst rendering the HTML for this, our server performs an internal fetch to another endpoint. We create a synthetic **REQ2** and **RES2** that this endpoint interacts with. It sets a cookie on **RES2**.
3. The internal fetch returns **RES2** and we extract the data, then render the rest of the HTML. Now we're ready to return a response to the user.

   We don't want to set _all_ cookies on this response from all requests we made as part of the rendering cycle. They could include logged in state from protected API calls we've made, for example, but it _is_ possible to set it manually. As I said previously:

   > If you want to allow the headers set on the response to useFetch to be added to the response from SSR, you'll have to handle that yourself. (You could even write your own composable to do this.)

I've created an [example reproduction](https://codesandbox.io/s/summer-cdn-1yx8s?file=/app.vue) with an example composable to show how you might 'pass on' a cookie from a local fetch to the final SSR response.
